packages/plugin-ext/src/plugin/timeline.ts

```diff
@@ -0,0 +1,138 @@
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { Plugin, TimelineCommandArg, TimelineExt, TimelineMain } from '../common';
+import { RPCProtocol } from '../common/rpc-protocol';
+import { Disposable, ThemeIcon } from './types-impl';
+import { PLUGIN_RPC_CONTEXT } from '../common';
+import { CancellationToken } from '@theia/core/lib/common/cancellation';
+import { DisposableCollection } from '@theia/core/lib/common/disposable';
+import { URI } from 'vscode-uri';
+import { PluginIconPath } from './plugin-icon-path';
+import { CommandRegistryImpl } from './command-registry';
+import * as theia from '@theia/plugin';
+import { Timeline, TimelineItem, TimelineOptions } from '@theia/timeline/lib/common/timeline-model';
+
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+// some code copied and modified from https://github.com/microsoft/vscode/blob/afacd2bdfe7060f09df9b9139521718915949757/src/vs/workbench/api/common/extHostTimeline.ts
+
+export class TimelineExtImpl implements TimelineExt {
+    private readonly proxy: TimelineMain;
+    private providers = new Map<string, theia.TimelineProvider>();
+    private plugin: Plugin;
+
+    private itemsBySourceAndUriMap = new Map<string, Map<string | undefined, Map<string, theia.TimelineItem>>>();
+
+    constructor(readonly rpc: RPCProtocol, private readonly commands: CommandRegistryImpl) {
+        this.proxy = rpc.getProxy(PLUGIN_RPC_CONTEXT.TIMELINE_MAIN);
+
+        commands.registerArgumentProcessor({
+            processArgument: arg => {
+                if (!TimelineCommandArg.is(arg)) {
+                    return arg;
+                }
+                if (arg.isSingleUri) {
+                    return URI.parse(arg.uri);
+                } else {
+                    return this.itemsBySourceAndUriMap.get(arg.source)?.get(arg.uri?.toString())?.get(arg.timelineHandle);
+                }
+            }
+        });
+    }
+
+    async $getTimeline(id: string, uri: string, options: TimelineOptions, internalOptions?: TimelineOptions): Promise<Timeline | undefined> {
```

Comment:
this does not code looks aligned with VS Code implementation, could you keep the plugin host code almost the same except some imports, are there any technical issues?

Reply:
Reworked the `timeline.ts` to be similar to the one from vscode as much as possible